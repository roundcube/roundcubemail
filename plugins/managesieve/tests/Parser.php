<?php

class Parser extends PHPUnit_Framework_TestCase
{

    function setUp()
    {
        include_once __DIR__ . '/../lib/Roundcube/rcube_sieve_script.php';
    }

    /**
     * Sieve script parsing
     *
     * @dataProvider data_parser
     */
    function test_parser($input, $output, $message)
    {
        // get capabilities list from the script
        $caps = array();
        if (preg_match('/require \[([a-z0-9", ]+)\]/', $input, $m)) {
            foreach (explode(',', $m[1]) as $cap) {
                $caps[] = trim($cap, '" ');
            }
        }

        $script = new rcube_sieve_script($input, $caps);
        $result = $script->as_text();

        $this->assertEquals(trim($result), trim($output), $message);
    }

    /**
     * Data provider for test_parser()
     */
    function data_parser()
    {
        $dir_path = realpath(__DIR__ . '/src');
        $dir      = opendir($dir_path);
        $result   = array();

        while ($file = readdir($dir)) {
            if (preg_match('/^[a-z0-9_]+$/', $file)) {
                $input = file_get_contents($dir_path . '/' . $file);

                if (file_exists($dir_path . '/' . $file . '.out')) {
                    $output = file_get_contents($dir_path . '/' . $file . '.out');
                }
                else {
                    $output = $input;
                }

                $result[] = array(
                    'input'   => $input,
                    'output'  => $output,
                    'message' => "Error in parsing '$file' file",
                );
            }
        }

        return $result;
    }
}
