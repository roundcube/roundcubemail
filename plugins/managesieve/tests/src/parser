require ["fileinto","reject","envelope"];
# rule:[spam]
if anyof (header :contains "X-DSPAM-Result" "Spam")
{
	fileinto "Spam";
	stop;
}
# rule:[test1]
if anyof (header :contains :comparator "i;ascii-casemap" ["From","To"] "test@domain.tld")
{
	discard;
	stop;
}
# rule:[test2]
if anyof (not header :contains :comparator "i;octet" ["Subject"] "[test]", header :contains "Subject" "[test2]")
{
	fileinto "test";
	stop;
}
# rule:[comments]
if anyof (true) /* comment
 * "comment" #comment */ {
    /* comment */ stop;
# comment
}
# rule:[reject]
if size :over 5000K {
	reject "Message over 5MB size limit. Please contact me before sending this.";
}
# rule:[false]
if false # size :over 5000K
{
	stop; /* rule disabled */
}
# rule:[true]
if true
{
	stop;
}
fileinto "Test";
# rule:[address test]
if address :all :is "From" "nagios@domain.tld"
{
	fileinto "domain.tld";
	stop;
}
# rule:[envelope test]
if envelope :domain :is "From" "domain.tld"
{
	fileinto "domain.tld";
	stop;
}
# rule:[do not wrap long lines]
if envelope :domain :is "From" "1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890"
{
	fileinto "domain.tld";
	stop;
}
