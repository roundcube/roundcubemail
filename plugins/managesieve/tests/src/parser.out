require ["envelope","fileinto","reject"];
# rule:[spam]
if header :contains "X-DSPAM-Result" "Spam"
{
	fileinto "Spam";
	stop;
}
# rule:[test1]
if header :contains ["From","To"] "test@domain.tld"
{
	discard;
	stop;
}
# rule:[test2]
if anyof (not header :contains :comparator "i;octet" "Subject" "[test]", header :contains "Subject" "[test2]")
{
	fileinto "test";
	stop;
}
# rule:[comments]
if true
{
	stop;
}
# rule:[reject]
if size :over 5000K
{
	reject "Message over 5MB size limit. Please contact me before sending this.";
}
# rule:[false]
if false # size :over 5000K
{
	stop;
}
# rule:[true]
if true
{
	stop;
}
fileinto "Test";
# rule:[address test]
if address :is "From" "nagios@domain.tld"
{
	fileinto "domain.tld";
	stop;
}
# rule:[envelope test]
if envelope :domain :is "From" "domain.tld"
{
	fileinto "domain.tld";
	stop;
}
# rule:[do not wrap long lines]
if envelope :domain :is "From" "1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890"
{
	fileinto "domain.tld";
	stop;
}
